### PR TITLE
Single Vue Instance

### PR DESCRIPTION
**Pull request by felixheidecke**
Fixes https://github.com/owncloud/phoenix/issues/94

## Things to be done:

- [x] Add vuex and vue-router to Phoenix base
- [x] move demo-app to expose a vue component
- [x] remove "old" helper and extend classes
- [x] Try to teach vue-router to dynamically extend routes
- [x] Use events instead of watchers to allow for single- and unwatching functionality
- [x] Move config loading to the store
- [x] Employ webpack to create smaller chunks
- [x] Add VueTidyRoutes to collect routes before instantiating vue
- [x] Fix resolve errors on chaning routes
- [x] Adapt files app to new structure
- [x] Fix `@TODO`s
- … _work in progress_

### CODE REVIEW

**Comment by felixheidecke** [comment]
Currently looking for a way to dynamically extend routes instead of defining them ahead of time.

**Comment by PVince81** [comment]
In general I think "ahead of time" should be acceptable, because: if you'd write a separate vue webapp, as far as I understand, you anyway need to define your routes upfront before booting the instance.
So here in Phoenix you could have a callback (extension point) from the main router where apps will register their routes.

**Comment by felixheidecke** [comment]
Not giving up dynamic route extending. Looking @ [vue-tidyroutes](https://github.com/edgarnadal/vue-tidyroutes) to move route models to their respective components. Let's see if this will do it :grimacing:

**Comment by felixheidecke** [comment]
The Idea is to load the app bundle from the server as it is needed, not ahead of time. The router needs to have the component model ready as they are beeing passed to vue upon instatiation. So I would be rather nice not to load all the bundles before mount :-|

I fear that I need to further separate bundles using webpack so I can it's async import function as it seems to work hand in hand with async vue components. Giving it a try.

**Comment by felixheidecke** [comment]
https://vuejs.org/v2/guide/components-dynamic-async.html#Async-Components

**Comment by PVince81** [comment]
> So I would be rather nice not to load all the bundles before mount :-|

makes sense

Unless we split the bundle to have a "routes.js" part containing only the route definitions upfront.

We have something similar in the backend where a separate PHP file contains only the routes. Not sure if that makes sense.

**Comment by felixheidecke** [comment]
Ha … I just had the best Idea ever. Event prevents ugly waiting and one less source for race conditions. Just wait a little bit.

**Comment by felixheidecke** [comment]
The current state implements a way that allows us to load apps and plugins easily using the webpack chunking. This makes the extra boot.js files obsolete and bundles are loaded directly whereas chunks are loaded on demand.

@OshanIvantha @PVince81 please check out and test using `config.json.sample` as your config file.
You can switch between applications in the menu and see a cloud plugin in the upper righthand corner.

If you like the approach, I'll start to adapt the current state of the files app as well as iron out a couple of "ugly" parts to be able to merge this.

**Comment by OshanIvantha** [comment]
@felixheidecke I get an error when I try to run this. 

screen shot 2018-06-28 at 12 01 53 am

File: apps/files/src/components/File-Details.vue [hunk 1]
```diff
@@ -84,7 +84,7 @@
 		},
 		watch : {
 			file () {
-				OC.$bus.emit('files:file-details-update', this.file)
+				OC.$events.emit('files:file-details-update', this.file)
```
Comment (PVince81) [inline apps/files/src/components/File-Details.vue]: do we still need these events ?

File: apps/servstat/src/app.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	computed : {
+		server () {
+			return this.$root.config.server;
+		}
+	}
+}
+
+const ip = {
+	name : "ServstatIp",
+	template : "<h1>Backend URL lautet {{ server }}</h1>",
```
Comment (PVince81) [inline apps/servstat/src/app.js]: sudden German ?

File: src/default.js [hunk 3]
```diff
@@ -1,288 +1,81 @@
...
+		}];
 
-		/**
-		 * Boot an application
-		 *
-		 * @param obj app with appId as key
-		 * @return Promise
-		 */
+		// @TODO: Need better implementation than [a, b, c, d, e, …]
+		_.forEach([a, b, c, d, e], (app) => {
```
Comment (PVince81) [inline src/default.js]: can you fix this right away ? at first glance it is very unclear what the letters are about

File: src/mixins/pluginHelper.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+		/**
+		 * Path to application file
+		 *
+		 * @param exp Extensionpoint
+		 * @param plugin Plugin
+		 */
+
+		getPlugin (ext) {
+			return _.find( this.$root.plugins, ['extend', ext]);
```
Comment (PVince81) [inline src/mixins/pluginHelper.js]: I wonder if this should be `this.$plugins` ? that would imply that accessing plugins after initialization is desired. probably not useful, so maybe keep it as is

File: apps/servstat/src/app.js (hunk 2 above)
Comment (felixheidecke) [reply apps/servstat/src/app.js]: Woops

File: apps/files/src/components/File-Details.vue (hunk 1 above)
Comment (felixheidecke) [reply apps/files/src/components/File-Details.vue]: Haven't touched the files app jet. (Except for some global find and replace stuff)

File: src/mixins/pluginHelper.js (hunk 4 above)
Comment (felixheidecke) [reply src/mixins/pluginHelper.js]: I'll try to prototype them, but right now these are given to the root element upon instantiation so in order to access them, we need to address `this.$root`.

**Comment by DeepDiver1975** [comment]
@felixheidecke conflicts

**Comment by individual-it** [comment]
I have the same problem as @OshanIvantha `127.0.0.1:35718 [404]: /apps/[object%20Object]/js/[object%20Object].bundle.js - No such file or directory` 
how did you fix it?

**Comment by OshanIvantha** [comment]
@individual-it 
Using only `make` doesn't work.
You need to run `npm install` and `npm run build` on *demo* and *serverstat* apps.

**Comment by individual-it** [comment]
@OshanIvantha then that might be the reason why the tests fail, need to adjust https://github.com/owncloud/phoenix/blob/master/.drone.yml#L60

**Comment by felixheidecke** [comment]
@individual-it please check out and test using config.json.sample as your config file.

**Comment by individual-it** [comment]
@felixheidecke OK got it running, but the UI changed quite a bit, so i will need to change the tests accordingly

**Comment by individual-it** [comment]
tests works fine locally but fail on drone. see https://drone.owncloud.com/owncloud/phoenix/54/3
any idea?

**Comment by individual-it** [comment]
OK test are modified and passing now

**Pull request merged**